File: webapi/src/main/java/dev/hekmyr/olids/api/dto/BillingInformationDTO.java
package dev.hekmyr.olids.api.dto;

import dev.hekmyr.olids.api.entity.BillingInformation;
import java.util.UUID;

public class BillingInformationDTO {

  private UUID id;
  private String cardNumber;
  private int monthExpiration;
  private int yearExpiration;
  private boolean isDefault;

  BillingInformationDTO() {}
  
  // Constructor made package-private or private if only factory method is intended
  BillingInformationDTO(BillingInformation entity) {
    this.id = entity.getId();
    this.cardNumber = entity.getCardNumber(); // Consider masking this number
    this.monthExpiration = entity.getMonthExpiration();
    this.yearExpiration = entity.getYearExpiration();
    this.isDefault = entity.isDefault();
  }

  public UUID getId() {
    return id;
  }

  public void setId(UUID id) {
    this.id = id;
  }

  public String getCardNumber() {
    return cardNumber;
  }

  public void setCardNumber(String cardNumber) {
    this.cardNumber = cardNumber;
  }

  public int getMonthExpiration() {
    return monthExpiration;
  }

  public void setMonthExpiration(int monthExpiration) {
    this.monthExpiration = monthExpiration;
  }

  public int getYearExpiration() {
    return yearExpiration;
  }

  public void setYearExpiration(int yearExpiration) {
    this.yearExpiration = yearExpiration;
  }

  public boolean isDefault() {
    return isDefault;
  }

  public void setDefault(boolean isDefault) {
    this.isDefault = isDefault;
  }

  public static BillingInformationDTO fromEntity(BillingInformation entity) {
    var dto = new BillingInformationDTO();
    dto.id = entity.getId();
    dto.cardNumber = entity.getCardNumber();
    dto.monthExpiration = entity.getMonthExpiration();
    dto.yearExpiration = entity.getYearExpiration();
    dto.isDefault = entity.isDefault();
    return dto;
  }
}

File: webapp/src/app/dto/billing-information.dto.ts
import { BillingInformation } from '../interface/billing-information.interface';

export class BillingInformationDTO {
  constructor(
    private _id: string,
    private _cardNumber: string,
    private _monthExpiration: number,
    private _yearExpiration: number
  ) {}

  public get id(): string {
    return this._id;
  }

  public get cardNumber(): string {
    return this._cardNumber;
  }

  public set cardNumber(value: string) {
    this._cardNumber = value;
  }

  public get monthExpiration(): number {
    return this._monthExpiration;
  }

  public set monthExpiration(value: number) {
    this._monthExpiration = value;
  }

  public get yearExpiration(): number {
    return this._yearExpiration;
  }

  public set yearExpiration(value: number) {
    this._yearExpiration = value;
  }

  public static fromInterface(
    billingInformation: BillingInformation
  ): BillingInformationDTO {
    return new BillingInformationDTO(
      billingInformation.id,
      billingInformation.cardNumber,
      billingInformation.monthExpiration,
      billingInformation.yearExpiration
    );
  }
}
